axlearn/cloud/gcp/job.py

```diff
@@ -639,6 +639,359 @@ def _execute(self) -> Any:
+                [
+                    {"name": "tcpx-socket", "mountPath": "/run/tcpx"},
+                    {"name": "nvidia-install-dir-host", "mountPath": "/usr/local/nvidia/lib64"},
+                    {"name": "tcpx-nccl-plugin-volume", "mountPath": "/usr/local/tcpx"},
+                ]
+            )
+            env_vars.update(
+                {
+                    "LD_LIBRARY_PATH": "/usr/local/tcpx/lib64:/usr/local/nvidia/lib64",
+                    "NCCL_CROSS_NIC": "0",
+                    "NCCL_ALGO": "Ring",
+                    "NCCL_PROTO": "Simple",
+                    "NCCL_DEBUG": "WARN",
+                    "NCCL_DEBUG_SUBSYS": "INIT,GRAPH,ENV,TUNING,NET,VERSION",
+                    "NCCL_NET_GDR_LEVEL": "PIX",
+                    "NCCL_P2P_PXN_LEVEL": "0",
+                    "NCCL_GPUDIRECTTCPX_FORCE_ACK": "0",
+                    "NCCL_GPUDIRECTTCPX_TX_COMPLETION_NANOSLEEP": "1000",
+                    "NCCL_GPUDIRECTTCPX_PROGRAM_FLOW_STEERING_WAIT_MICROS": "1000000",
+                    "NCCL_GPUDIRECTTCPX_TX_BINDINGS": (
+                        "eth1:8-21,112-125;eth2:8-21,112-125;"
+                        "eth3:60-73,164-177;eth4:60-73,164-177"
+                    ),
+                    "NCCL_GPUDIRECTTCPX_RX_BINDINGS": (
+                        "eth1:22-35,124-139;eth2:22-35,124-139;"
+                        "eth3:74-87,178-191;eth4:74-87,178-191"
+                    ),
+                    "NCCL_GPUDIRECTTCPX_SOCKET_IFNAME": "eth1,eth2,eth3,eth4",
+                    "NCCL_GPUDIRECTTCPX_CTRL_DEV": "eth0",
+                    "NCCL_GPUDIRECTTCPX_UNIX_CLIENT_PREFIX": "/run/tcpx",
+                    "NCCL_DYNAMIC_CHUNK_SIZE": "524288",
+                    "NCCL_P2P_NET_CHUNKSIZE": "524288",
+                    "NCCL_P2P_PCI_CHUNKSIZE": "524288",
+                    "NCCL_P2P_NVL_CHUNKSIZE": "1048576",
+                    "NCCL_NSOCKS_PERTHREAD": "4",
+                    "NCCL_SOCKET_NTHREADS": "1",
+                    "NCCL_SOCKET_IFNAME": "eth0",
+                    "NCCL_NVLS_ENABLE": "0",
+                }
+            )
+
+        # Override env vars with user provided env vars
+        env_vars.update(cfg.env_vars)
+        # K8s expects each env variable to be a dict
+        k8s_env_vars = [{"name": name, "value": value} for name, value in env_vars.items()]
+        k8s_env_vars.append(
+            {
+                "name": "PROCESS_ID",
+                "valueFrom": {
+                    "fieldRef": {
+                        "fieldPath": (
+                            "metadata.annotations['batch.kubernetes.io/" "job-completion-index']"
+                        ),
+                    }
+                },
+            },
+        )
+
+        user_cmd = cfg.command
+        # This is needed to make the sidecar exit when the main container exits.
+        if cfg.accelerator.instance_type.startswith("gpu-a3-highgpu"):
+            user_cmd += "\ntouch /run/tcpx/terminated"
```

Comment:
Yes it's expected. New line works like a semicolon in this case. I verified this is working. Let me know if you prefer a semicolon though, because that should work too. I can retest with a semicolon if that's your preference.

Reply:
fixed